queue/group.go

```diff
@@ -158,3 +177,360 @@ func (g *localQueueGroup) Close(ctx context.Context) {
 		return
 	}
 }
+
+// remoteQueueGroup is a group of database-backed queues.
+type remoteQueueGroup struct {
+	canceler       context.CancelFunc
+	client         *mongo.Client
+	constructor    RemoteConstructor
+	mongooptions   MongoDBOptions
+	mu             sync.RWMutex
+	prefix         string
+	pruneFrequency time.Duration
+	queues         map[string]amboy.Queue
+	ttl            time.Duration
+	ttlMap         map[string]time.Time
+}
+
+type RemoteQueueGroupOptions struct {
+	// Client is a connection to a database.
+	Client *mongo.Client
+
+	// Constructor is a function passed by the client to construct a remote queue.
+	Constructor RemoteConstructor
+
+	// MongoOptions are connection options for the database.
+	MongoOptions MongoDBOptions
+
+	// Prefix is a string prepended to the queue collections.
+	Prefix string
+
+	// PruneFrequency is how often Prune runs.
+	PruneFrequency time.Duration
+
+	// TTL is how old the oldest task in the queue must be for the collection to be pruned.
+	TTL time.Duration
+}
+
+type listCollectionsOutput struct {
+	Name string `bson:"name"`
+}
+
+// NewRemoteQueueGroup constructs a new remote queue group. If ttl is 0, the queues will not be
+// TTLed except when the client explicitly calls Prune.
+func NewRemoteQueueGroup(ctx context.Context, opts RemoteQueueGroupOptions) (amboy.QueueGroup, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	if err := opts.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid remote queue options")
+	}
+
+	colls, err := getExistingCollections(ctx, opts.Client, opts.MongoOptions.DB, opts.Prefix)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem getting existing collections")
+	}
+
+	g := &remoteQueueGroup{
+		canceler:       cancel,
+		client:         opts.Client,
+		constructor:    opts.Constructor,
+		mongooptions:   opts.MongoOptions,
+		prefix:         opts.Prefix,
+		pruneFrequency: opts.PruneFrequency,
+		ttl:            opts.TTL,
+	}
+
+	g.queues = map[string]amboy.Queue{}
+	g.ttlMap = map[string]time.Time{}
+	catcher := grip.NewBasicCatcher()
+	for _, coll := range colls {
+		q, err := g.startProcessingRemoteQueue(ctx, coll)
+		if err != nil {
+			catcher.Add(errors.Wrap(err, "problem starting queue"))
+		} else {
+			g.queues[idFromCollection(g.prefix, coll)] = q
+			g.ttlMap[idFromCollection(g.prefix, coll)] = time.Now()
+		}
+	}
+	if catcher.HasErrors() {
+		return nil, catcher.Resolve()
+	}
+
+	if opts.PruneFrequency > 0 && opts.TTL > 0 {
+		go func() {
+			defer recovery.LogStackTraceAndContinue("panic in remote queue group ticker")
+			ticker := time.NewTicker(opts.PruneFrequency)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					g.Prune(ctx)
+				}
+			}
+		}()
+	}
+	return g, nil
+}
+
+func (g *remoteQueueGroup) startProcessingRemoteQueue(ctx context.Context, coll string) (Remote, error) {
```

Comment:
there should be a way of constructing one of these that doesn't require it's own connection pool